Dark Mode toggle

zackkrida opened the pull request:
> # :warning: Based on the Ui Updates PR #4380 

<!-- Concisely describe what the pull request does. -->

This PR adds a toggle to Mathesar's header which allows the user to choose between "Dark", "Light" and "System" themes, with the later inferring the correct theme from the user's operating system level preference. The later choice is also the default.

@seancolsen, please review the app itself to ensure the correct functionality and the following files for code quality:

- mathesar_ui/src/App.svelte
- mathesar_ui/src/AppContext.svelte
- mathesar_ui/src/components/AppHeader.svelte
- mathesar_ui/src/stores/localStorage.ts
- mathesar_ui/src/utils/uiThemePreference.ts

@ghislaineguerin, please review the app itself to ensure the correct functionality, and review the styles, in particular, to make sure that I didn't break any styles in converting from the media query to theme class method for applying dark mode styles.

**Technical details**
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->

- Adds a theme switcher UI component in the AppHeader dropdown 
- Stores the selected theme in a new localStorage store
- Applies the theme by setting a `body.theme-{light|dark}` class  
- Replaces all `prefers-color-scheme: dark` media queries with `body.theme-dark` selectors in our components  
- Add initialization logic to apply the selected theme on app load and listen to changes in the user's system-level preference

**Screenshots**
<!-- Add screenshots to show the problem and the solution; or delete the section entirely. -->

![Screenshot From 2025-04-10 17-42-50](https://github.com/user-attachments/assets/28422939-6999-4a6a-b985-3852ef1eead3)
![image](https://github.com/user-attachments/assets/81b3bd78-4dcb-4af1-8378-cfb8658b731e)


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

zackkrida commented on mathesar_ui/src/components/AppHeader.svelte:
```diff
@@ -76,7 +77,7 @@
           triggerAppearance="ghost"
           size="small"
-          closeOnInnerClick={true}
+          closeOnInnerClick={false}
```
This change is so the user can interact with the theme switcher. It is only noticed when choosing the documentation links, as internal link clicks already close the menu by means of page navigation.

seancolsen reviewed (changes requested):
Aside from the tests failures, the only thing that I think really needs to be changed here is the `on:click` issue.

- You have a TS error, which I think is related to the `on:click` change I mentioned.

- There are some ESLint errors too, that I'll let you fix.

- The code in `uiThemePreference.ts` seems sane, but took me a little while to grok, just because it was so different from how I would have done it. 

    I'll describe how I would have done it, just in case this is useful for you in getting more familiar with Svelte:

    - I'd use a stores-based approach with the logic in `src/stores/theme.ts` where I'd:
        - Use a simpler type:

            ```ts
            export type UiTheme = 'light' | 'dark'; // No 'system' option
            ```

        - Have a writable `themePreference` store which is `Writable<UiTheme | undefined>`;
        - Have a readable `systemTheme` store which is `Readable<UiTheme>`;
        - Have a derived `theme` store which combines the two above stores with nullish coalescing. It would then be `Readable<UiTheme>`.

    - Within AppContext.svelte, I'd read the `theme` store to set the class on `<svelte:body />` declaratively/reactively.

    I don't think it's worth changing your code since it appears to work. Just food for thought next time!

zackkrida commented:
@seancolsen thanks for the review. I've fixed the lint and type issues (there are still some from the base branch), removed the unnecessary window check (a habit from writing a lot of isomorphic code in server-rendered contexts), and re-added the click handler to the menu wrapper items. This is all in e724fb6.

As for the imperative code in uiThemePreference, this gets interesting! `<svelte:body />` can _only_ be used to register global event handlers—not to set a CSS class on the body. That limitation, combined with the existing pattern of colocating all LocalStorageStore instances in a single file, led me to the current implementation. I thought it made more sense to keep all the more imperative code relating to the UI theme in one file rather than some in the AppContext and some in the store.

seancolsen reviewed (approved):
Thanks @zackkrida. Looks great!

> `<svelte:body />` can _only_ be used to register global event handlers

Oh, interesting! I didn't know that. TIL! Thanks! I can see how that would tip the scale towards more imperative code. Nonetheless, I think I'd still be inclined to use the writable-readable-derived stores approach, but it's probably more of a personal preference than a code design/standards issue in this case.

seancolsen commented:
Separate from my review, I'll say: I'm not really sure how best to handle this PR, process-wise.

- **Merge it** into Ghislaine's PR? That's not what we typically do with stacked PRs, because it _usually_ complicates review of the base PR. We'd want to avoid re-reviewing the code changes here. But I think in this case the theme switcher actually makes it quite a bit easier to review the base PR because it's quicker to toggle the theme setting.

- **Hold it** open until Ghislaine's PR is merged? That's what we usually do. I'm not sure how long we'd be waiting in this case though, given that the base PR is still in draft state. Managing long-lived stacked PRs can become a little cumbersome.

@zackkrida I'll leave the choice to you. I'm fine with either. I just wanted to voice some of my thinking here.

zackkrida commented:
I appreciate your insights on the PR process @seancolsen. I definitely think in this case we should merge it into the base PR because it adds a new paradigm for defining dark mode CSS (using the `.dark-mode` class rather than a media query). 

I'll go ahead and merge it now. Thanks!
